### ISSUE TITLE
Bug when using dtls?

### ISSUE THREAD

**Issue by TethysSvensson** [namib-project/libcoap-rs#7]
It would be nice if there was some example for how to use dtls correctly. I am trying to use this code, but it panics with `unable to restore CoapLendableFfiRcCell as the underlying value was already dropped`.

I am unable to figure out if this is a bug in my code, or in libcoap-rs.

```rust
use std::net::SocketAddr;

use libcoap_rs::{
    crypto::{CoapClientCryptoProvider, CoapCryptoProviderResponse, CoapCryptoPskInfo},
    session::CoapClientSession,
    CoapContext,
};

#[derive(Debug)]
struct MyProvider;

impl CoapClientCryptoProvider for MyProvider {
    fn provide_key_for_hint(
        &mut self,
        _hint: &libcoap_rs::crypto::CoapCryptoPskIdentity,
    ) -> CoapCryptoProviderResponse<Box<libcoap_rs::crypto::CoapCryptoPskData>> {
        CoapCryptoProviderResponse::UseCurrent
    }

    fn provide_default_info(&mut self) -> CoapCryptoPskInfo {
        CoapCryptoPskInfo {
            identity: (*b"myClient").into(),
            key: (*b"mysecretKey").into(),
        }
    }
}

fn main() {
    let server_address: SocketAddr = "127.0.0.1:5684".parse().unwrap();

    let mut context = CoapContext::new().expect("Failed to create CoAP context");

    let session = CoapClientSession::connect_dtls(&mut context, server_address, MyProvider)
        .expect("Failed to create client-side session");

    std::mem::drop(session);
    std::mem::drop(context);
}
```

**Comment by TethysSvensson** [comment]
Specifically this crashes inside `event_handler_callback` in `event.rs`, which is called with the event `COAP_EVENT_DTLS_CLOSED`. It crashes when it tries to call `CoapContext::from_raw`.

**Comment by TethysSvensson** [comment]
If I add this `Drop` impl for `CoapContext`, then it seems to work:

```rust
impl Drop for CoapContext<'_> {
    fn drop(&mut self) {
        let mut inner_ref = self.inner.borrow_mut();
        let client_sessions = std::mem::take(&mut inner_ref.client_sessions);
        let server_sessions = std::mem::take(&mut inner_ref.server_sessions);
        std::mem::drop(inner_ref);

        for session in client_sessions {
            session.drop_exclusively();
        }
        for session in server_sessions {
            session.drop_exclusively();
        }
    }
}
```

I don't know the library well enough to know if this is a good idea, but I can make a PR if you like.

**Comment by pulsastrix** [comment]
Thanks for the report!

I have found the underlying issue: We did not disable event handling before dropping the context and its' sessions. 

Dropping a DTLS session triggers a `COAP_EVENT_DTLS_CLOSED` event, but the event handler requires an existing reference to the context to be provided to it. As we don't provide such a reference, the event handler panics.

I've started work on a fix in #8 and will probably get a new release ready in the next few days.
